pkg/webhook/cluster/mutating_webhook.go

```diff
@@ -58,18 +67,115 @@ func (a *ManagedClusterMutatingAdmissionHook) Admit(req *admissionv1beta1.Admiss
 		return status
 	}
 
+	var jsonPatches []jsonPatchOperation
+
 	// If LeaseDurationSeconds value is zero, update it to 60 by default
 	if managedCluster.Spec.LeaseDurationSeconds == 0 {
-		status.Patch = []byte(defaultLeaseDurationSecondsPatch)
-		pt := admissionv1beta1.PatchTypeJSONPatch
-		status.PatchType = &pt
+		jsonPatches = append(jsonPatches, newLeaseDurationSecondsJsonPatch(60))
+	}
+
+	// set timeAdded of taint if it is nil and reset it if it is modified
+	taintJsonPatches, err := a.processTaints(managedCluster, req.OldObject.Raw)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusInternalServerError, Reason: metav1.StatusReasonInternalError,
+			Message: err.Error(),
+		}
+		return status
 	}
+	jsonPatches = append(jsonPatches, taintJsonPatches...)
+
+	if len(jsonPatches) == 0 {
+		return status
+	}
+
+	patch, err := json.Marshal(jsonPatches)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusInternalServerError, Reason: metav1.StatusReasonInternalError,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	status.Patch = patch
+	pt := admissionv1beta1.PatchTypeJSONPatch
+	status.PatchType = &pt
 
 	return status
 }
 
+// processTaints generates json patched for cluster taints
+func (a *ManagedClusterMutatingAdmissionHook) processTaints(managedCluster *clusterv1.ManagedCluster, oldManagedClusterRaw []byte) ([]jsonPatchOperation, error) {
+	if len(managedCluster.Spec.Taints) == 0 {
+		return nil, nil
+	}
+
+	var oldManagedCluster *clusterv1.ManagedCluster
+	if len(oldManagedClusterRaw) > 0 {
+		oldManagedCluster = &clusterv1.ManagedCluster{}
+		if err := json.Unmarshal(oldManagedClusterRaw, oldManagedCluster); err != nil {
+			return nil, err
+		}
+	}
+
+	var jsonPatches []jsonPatchOperation
+	now := metav1.NewTime(nowFunc())
+	for index, taint := range managedCluster.Spec.Taints {
+		// reuse the existing value of timeAdded if all the key/value/effect are matched
+		// otherwise reset the value
+		timeAdded := getTaintTimeAdded(oldManagedCluster, taint.Key, taint.Value, taint.Effect, now)
```

Comment:
I think we just need to set timeAdded if it is nil? why do we need to compare it with the old value. To make it immutable?

Reply:
Yes. After a taint is created, the timeAdded should be immutable unless its key/value/effect changes. It should be maintained by the webhook rather than user.

Reply:
i think we need validatingwebhook to do the immutable check rather than implicitly mutate it

Reply:
Validation has been added in the mutating webhook instead of validatingwebhook.